packages/replay-next/src/suspense/AnalysisCache.ts

```diff
@@ -68,31 +72,38 @@ function getCacheKey(params: AnalysisParams) {
   return cacheKey + params.mapper;
 }
 
-export function getAnalysisCache<T extends { point: ExecutionPoint }>(
-  params: AnalysisParams,
-  transformPoint: (point: PointDescription) => T
-): AnalysisCache<T> {
+const analysisResultsCaches = new Map<string, Map<ExecutionPoint, Record<RemoteAnalysisResult>>>();
+function getAnalysisResultsCache(params: AnalysisParams) {
```

Comment:
So this function is basically responsible for creating two, synchronized caches: an interval cache (for fetching points/hits) and a regular cache (for evaluating data at a point).

I think we could remove a bit of redundancy here by using `createCache` for the second cache. What do you think?

```ts
export function createAnalysisCache<
  TPoint extends { point: ExecutionPoint },
  TParams extends any[]
>(
  debugLabel: string,
  createAnalysisParams: (...params: TParams) => AnalysisParams,
  transformPoint: (point: PointDescription, ...params: TParams) => TPoint
): AnalysisCache<TPoint, TParams> {
  const resultsCache: Cache<
    [executionPoint: ExecutionPoint, ...params: TParams],
    RemoteAnalysisResult
  > = createCache({
    debugLabel: `${debugLabel} Cache`,
    getKey: ([executionPoint, ...params]) => {
      const analysisParams = createAnalysisParams(...params);
      const baseKey = getCacheKey(analysisParams);
      return `${baseKey}:${executionPoint}`;
    },
    load: async ([executionPoint, ...params]) => {
      // This cache doesn't load its own values;
      // Values are externally cached by the interval cache.
      return new Promise((resolve, reject) => {
        setTimeout(() => {
          reject(
            "Analysis results cache timed out; this may indicate that the points interval cache was not loaded."
          );
        }, 5_000);
      });
    },
  });

  const pointsIntervalCache = createIntervalCache<
    ExecutionPoint,
    [client: ReplayClientInterface, ...params: TParams],
    TPoint
  >({
    debugLabel: `${debugLabel} IntervalCache`,
    getKey: (client, ...params) => getCacheKey(createAnalysisParams(...params)),
    getPointForValue: pointDescription => pointDescription.point,
    comparePoints: compareNumericStrings,
    load: async (begin, end, client, ...paramsWithCacheLoadOptions) => {
      const params = paramsWithCacheLoadOptions.slice(0, -1) as TParams;
      const analysisParams = createAnalysisParams(...params);
      const locations = analysisParams.location
        ? client.getCorrespondingLocations(analysisParams.location).map(location => ({
            location,
          }))
        : undefined;
      if (locations) {
        await Promise.all(
          locations.map(location =>
            breakpointPositionsCache.readAsync(client, location.location.sourceId)
          )
        );
      }

      let allPoints: TPoint[] = [];
      let error: any;
      await client.streamAnalysis(
        {
          locations,
          eventHandlerEntryPoints: analysisParams.eventTypes?.map(eventType => ({ eventType })),
          exceptionPoints: analysisParams.exceptions,
          mapper: analysisParams.mapper,
          effectful: false,
          range: { begin, end },
        },
        {
          onPoints: points => {
            allPoints = allPoints.concat(points.map(point => transformPoint(point, ...params)));
          },
          onResults: analysisEntries => {
            for (const analysisEntry of analysisEntries) {
              const result = analysisEntry.value as RemoteAnalysisResult;
              cachePauseData(client, result.pauseId, result.data);

              // This will resolve any pending/in-flight request from the Cache
              resultsCache.cache(result, result.point, ...params);
            }
          },
          onError: err => (error = err),
        }
      ).pointsFinished;

      if (error) {
        throw error;
      }
      if (allPoints.length > MAX_POINTS_FOR_FULL_ANALYSIS) {
        throw new Error("Too many points to run analysis");
      }
      return allPoints;
    },
  });

  return {
    pointsIntervalCache,
    resultsCache,
  };
}
```

Reply:
I kind of like that the inner cache will time out with an explicit error if we "mess up" (especially since it's a non-standard way of using the caches)

Reply:
Note #8973 upgrades `suspense@0.0.29` with a small change I just made to ensure that calling `cache` will clear out/cancel any pending requests (so the above suggestion will work).

Reply:
I don't have a strong opinion here although I lean towards keeping it the way it is because using `createCache` with a dummy `load` method feels like we're overusing it.

Reply:
I prefer a more explicit coupling (with a timeout that will throw ratner than just permanently hang). I'd also like to avoid avoiding duplicating a chunk of the cache API (like the PR is currently doing).

I agree that the way `load` is used is a bit odd. Maybe a special, explicit "externally managed cache" type could alleviate that concern? It would basically be the same thing but its config wouldn't require a `load` method and could instead accept an optional `timeout` param?

```ts
export function createExternallyManagedCache<Params extends Array<any>, Value>(
  options: Omit<CreateCacheOptions<Params, Value>, "load"> & {
    timeout?: number;
    timeoutMessage?: string;
  }
): Cache<Params, Value> {
  const { timeout = 5_000, timeoutMessage = "Timed out", ...rest } = options;

  return createCache<Params, Value>({
    ...rest,
    load: async ([executionPoint, ...params]) => {
      // This cache doesn't load its own values;
      // Values are externally cached by the interval cache.
      return new Promise((resolve, reject) => {
        setTimeout(() => {
          reject(timeoutMessage);
        }, timeout);
      });
    },
  })
}
```

Reply:
I think part of the reason I prefer this is that there are fewer code paths– so if we have test coverage of one path, or fix a bug in one path, all paths benefit.

It's also nice to have actual `Cache` objects because they can be used with the other "suspense" hooks where as a one-off faux-cache can't be.